# EV Charging Patterns — Clustering Analysis
**Dataset:** [EV Charging Patterns (Kaggle)](https://www.kaggle.com/datasets/valakhorasani/electric-vehicle-charging-patterns)  
**Goal:** Discover meaningful clusters in EV charging sessions using unsupervised learning.

### Phase 1 Data Understand & Preparation
#### 1.1 Exploration

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
import warnings
warnings.filterwarnings('ignore')

plt.rcParams.update({'figure.figsize':(12, 8), 'axes.titlesize': 14, 'axes.labelsize':12})

In [4]:
df = pd.read_csv('ev_charging_dataset/ev_charging_patterns.csv')

In [9]:
print("Dataset Overview")
print(f" Shape: {df.shape}")

print("\nData Types")
print(df.dtypes)

print("\nMissing Values")
print(df.isnull().sum())

Dataset Overview
 Shape: (1320, 20)

Data Types
User ID                                      object
Vehicle Model                                object
Battery Capacity (kWh)                      float64
Charging Station ID                          object
Charging Station Location                    object
Charging Start Time                          object
Charging End Time                            object
Energy Consumed (kWh)                       float64
Charging Duration (hours)                   float64
Charging Rate (kW)                          float64
Charging Cost (USD)                         float64
Time of Day                                  object
Day of Week                                  object
State of Charge (Start %)                   float64
State of Charge (End %)                     float64
Distance Driven (since last charge) (km)    float64
Temperature (°C)                            float64
Vehicle Age (years)                         float64
Charger Type    

**Reflection 1.1** The dataset contains 1320 charging sessions with 20 features. Things like vehicle specs, session details and context(what time of day for instance) are covered. There are missing values in the following 3 columns: *Energy Consumed kWh*, *Charging Rate kW*, and
*Distance Driven since last charge km* (66 rows each).

### 1.2 Handling Missing Data & Basic Statistics
In order to deal with the missing data, they get replaced with median value. That way, its robust to outliers. if we would have picked the average, and a certain value was inflated because of a charging session that artificially took a long time, it would effect the average when its an outlier. So imputation is used. imputation is when the missing value is replaced with the median value.

In [14]:
print("Missing values  before imputation")
print(df.isnull().sum())

numeric_imputer = SimpleImputer(strategy='median')
cols_with_missing = ['Energy Consumed (kWh)',
    'Charging Rate (kW)',
    'Distance Driven (since last charge) (km)']

df[cols_with_missing] = numeric_imputer.fit_transform(df[cols_with_missing])

Missing values  before imputation
User ID                                      0
Vehicle Model                                0
Battery Capacity (kWh)                       0
Charging Station ID                          0
Charging Station Location                    0
Charging Start Time                          0
Charging End Time                            0
Energy Consumed (kWh)                       66
Charging Duration (hours)                    0
Charging Rate (kW)                          66
Charging Cost (USD)                          0
Time of Day                                  0
Day of Week                                  0
State of Charge (Start %)                    0
State of Charge (End %)                      0
Distance Driven (since last charge) (km)    66
Temperature (°C)                             0
Vehicle Age (years)                          0
Charger Type                                 0
User Type                                    0
dtype: int64
